### PR TITLE
Improvements to MOSEK Direct interface and a new persistent interface

### LINKED ISSUE #1513: Efficiency of MOSEK interface
The MOSEK direct interface inputs variables and constraints into the MOSEK task object one by one. The generic method ``_add_block`` inherited from the abstract direct solver calls the methods ``_add_var`` and ``_add_constraint`` for each variable/constraint. It is not particularly efficient, and especially in MOSEK it would be much faster to input at least the whole linear data in one go. 

Here at MOSEK we wanted to see if we could achieve something like that, for instance by redefining ``_add_block`` in the MOSEK interface. So I wanted to ask if this is not against some design principle (as far as I can see all the direct interfaces follow the one-by-one pattern), if you have some suggestions how one would go about it, or if there already are any plans of this kind.

### PR DESCRIPTION
**Pull request by Utkarsh-Detha**
## Fixes #1513 

# Motivation:
While investigating why the direct interface was slow to pass the problem to the solver (see google group [discussion](https://groups.google.com/g/pyomo-forum/c/PE2GV0MGBD8)), @aszekMosek found that `_add_var` and `_add_constraint` methods pass one variable/constraint to the MOSEK task per call, thereby making it a bit slow. Furthermore, @michaelbynum stated that this was a design choice made to accommodate persistent interfaces. This pull request is therefore motivated by a need to add some performance improvements to the MOSEK direct interface and making a new persistent interface for MOSEK.

# Major changes proposed in this PR:
1. `add_vars` and `add_constraints` methods are included. These take a list of variables/constraints (single data objects), extract the necessary information to be passed to the MOSEK task object, and make one call to the appropriate MOSEK Optimizer API methods. Naturally, `_add_block` has been overridden to use these methods.
2. `_add_var` and `_add_constraint` methods are now simply calls to the `add_vars` and `add_constraints` methods.
3. A persistent interface defined as the class MOSEKPersistent and registered in the SolverFactory as `"mosek_persistent"` is now available. Methods are available to remove variable(s) and constraint(s), modify variable(s), add a column etc.
4. Included tests for the persistent interface in test_MOSEKPersistent.py.

### Minor changes proposed in the PR:
1. `MosekDirect `class is now called `MOSEKDirect` and the direct interface is registered in the SolverFactory as  `"mosek_direct"`.
2. `apply_solver` method in the direct interface now prints a termination code in case of a significant failure.
3. `_get_cone_data` method always sets the continuity relaxation to True.
4. `remove_vars`, `remove_constraints` and `update_vars` methods in the persistent interface will accept multiple singleton var/constraint objects, passed as individual arguments. Analogous methods that accept a single object are also available.
5. test_mosek.py is now called test_MOSEKDirect.py

## Benchmarks (direct interface) [updated]
A performance improvement of about ~30 % was observed across the following benchmarks: 
1. (Small) Issue #51: Averaging over 10 runs, the current master branch took 11.96 seconds, while this PR took 6.79 seconds.
2. (Medium) A model to calculate Wasserstein distance, _using n=2 images_ (see the linear model provided [here](https://themosekblog.blogspot.com/2019/05/mosek-in-pyomo-and-wasserstein-distances.html)): the current master branch took 75.14 seconds while the PR took 55.1 seconds. Solver time in both cases was around 1.5 seconds.
3. (Large) A model to calculate Wasserstein distance, _using n=20 images_ (see the linear model provided [here](https://themosekblog.blogspot.com/2019/05/mosek-in-pyomo-and-wasserstein-distances.html)): the current master branch took about 803 seconds while this PR took 581 seconds. Solver time in both cases was about 40 seconds.

NOTE: Suggestions made by @jsiirola further improved the performance of the interface, and the benchmark results provided above were updated accordingly.

I ran the benchmarks on an Intel i7 10875h, with 32 GB RAM.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by Utkarsh-Detha** [comment]
Could someone (maybe @jsiirola) give me a hint as to why the Jenkins CI /inspection has not started yet? I am still new to the world of CI/CD and any help would be greatly appreciated.

**Comment by blnicho** [comment]
@Utkarsh-Detha someone on the core development team needs to do a preinspection on the PR before the Jenkins tests are launched. Our weekly developers meeting is on Tuesday so you should see some movement on this PR later today.

**Comment by Utkarsh-Detha** [comment]
Thanks for your response, @blnicho ! One more thing: I see that the codecov/patch test has failed and the culprit seems to be the foqus_graph.py file (I say that simply because the delta is negative only in that file's case). However, I have not tampered with this file, and therefore do not understand how this could happen. Any advice would be much appreciated.

**Comment by jsiirola** [comment]
@Utkarsh-Detha, You always have to take codecov results with a  grain of salt: there are frequently inexplicable changes in coverage that are unrelated to the PR.  That said, `codecov/patch` fails when the PR patch (the set of changed lines) is covered at a level that is less than the overall project.  In this case, that is because we do not (yet) have Mosek available on any of our testing platforms (as it requires a license, we cannot set it up on the public platforms and we have been having challenges getting approvals to provide the necessary information to obtain a license for our nonpublic testing platform).

**Comment by aszekMosek** [comment]
Regarding testing you can use http://solve.mosek.com if you set it up appropriately: locally you would only need the interface, which is not licensed, and you would somehow have to hack the Pyomo interface to enable remote optimization.

See for example what happens in Julia:

https://github.com/MOSEK/Mosek.jl/blob/112e0d7d7f8bb753360abc3c3693f9dccd5c6fb7/test/apitest.jl#L43

**Comment by Utkarsh-Detha** [comment]
@jsiirola am I right in thinking that the PR will not be considered for merging as it fails the codecov/patch test? If so, I can close this PR and open another one with @aszekMosek's suggestion incorporated in it. How would you suggest we proceed?

**Comment by jsiirola** [comment]
No - I do not think this PR needs to be closed, although an extension that supports the remote interface that @aszekMosek highlights would be great.

(I am most of the way through my review, so hopefully I can get that finished and posted soon)

**Review by jsiirola — CHANGES_REQUESTED** [review]
This is a good start.  Several questions below.

**Comment by jsiirola** [comment]
@Utkarsh-Detha, please bundle your changes into fewer pushes to github.  Each push (or direct edit on github) to a branch with an open PR triggers the GHA tests, which in turn launches about 20 jobs that can run upwards of a half an hour.  The recent individual file edits created a significant backlog in the testing infrastructure that impacted other PRs / development activity.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:21]: These imports are actually identical to the older interface and the Cplex/Gurobi direct interface. I simply re-arranged the lines.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py]: I had this concern myself and was planning to make a follow-up PR sometime soon with the said "meta" solver. In the meantime, I will admit, I was hoping that this would not be a huge deal-breaker.

**Comment by Utkarsh-Detha** [comment]
Ah, I am sorry for the inconvenience. I will do as you suggested.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:68]: The exponential and power cones are supported only from version 9 onwards. So, the major version number is cached to be able to perform that check later in the code.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:99]: I never thought about this before, but it is in fact more efficient to use the checkoutlicense methods! I ran a loop checking license availability 10**4 times, and found that if a license is available, then `optimze()` method will take about 69 seconds (84% CPU usage) and the `checkoutlicense() `method method will take about 58 seconds (with 94% CPU usage). I will make the necessary changes in the file.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:139]: I do this because integer parameters accept enumerations, such as `mosek.onoffkey.on`  (see [here](https://docs.mosek.com/9.2/pythonapi/parameters.html#integer-parameters)). So my rationale was that if a user reads about the accepted values to a parameter from MOSEK's online documentation, and decides to use them verbatim, they should be able to.
I don't check for it in the other types of parameters because they accept either doubles or strings.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:191]: You are right, my bad. I will re-introduce `value()`.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:218]: I remember using the update method before, but replaced it with for some reason which escapes me now. No matter, I will use the update method. Thank you!

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:57]: Will make the change, thanks!

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:128]: Do you mean that I should replace the print statement with something like this: `logger.info('Solver log file : ' + self._log_file)`?
If so, then I am confused about the purpose of this statement: are we not trying to provide the user with some sort of confirmation that the log will be saved in a file that has the name stored in `self._log_file`?

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:235]: Agreed.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:330]: Could you suggest an alternative to checking this attribute? I have checked in the old MOSEK interface and the CPLEX interface for instance, and they both seem to be using this attribute.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:359]: I am doing this because, for instance, `l_ids` is a list of lists, and it needs to be flattened before passing to the `putarowlist()` method.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py]: The old interface used eval, and I just typed that in myself, without thinking of the better option that you have suggested. I will make the change. Thanks!

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py]: I agree with your comment about the implication of the error message. I will now _log_ an error message that simply states that the Task creation failed and then append a string that describes the error itself. This should also help the user pin-point the actual reason for the said failure.

Catching `Exception` and not using the bare `except` was what the old pyomo-mosek interface had. I left it unchanged. I will do it now.

About catching a generic exception: fair point! I will log and re-raise the exception. Thank you!

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (jsiirola) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:128]: Yes.  I am proposing
```python
logger.info("Solver log file: %s" % (self._log_file,))
```
The `logger` object would be from something like
```python
import logging
logger = logging.getLogger(__name__)
```
This is a *python* logger and not the logger used by the solver (and written to `self._log_file`).  So the information should be available to the user.  The problem with `print()` is that it makes it harder to embed Pyomo in larger applications.  I realize that the `print` is from the original implementation, but since we are cleaning up the interface, I thought I would bring it up.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py]: The last case that you mention, i.e. 1< var <2 seems correct to me. For the other two cases, MOSEK will raise an error due to invalid input: putvarbound method will not accept `float('inf')` as a valid input. Furthermore, if a user inputs an upper bound that is smaller than a lower bound (both finite), then it will make the problem Primal infeasible, which will be evident in the log output. So there should be no cases where the problem would give wrong results due to this.

However, as you mentioned in one of your other comments, if this makes it harder to implement pyomo in other appilcations, then I will change this to the method that was used before.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:128]: Understood, I will make the change. Thank you!

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (jsiirola) [reply pyomo/solvers/plugins/solvers/mosek_direct.py]: I am not sure the previous implementation is better.  My only intent with the comment was there was an weird edge case and it took me longer to figure out what the 1-liner `bound_key = (...` was doing than I would like to admit.

Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py]: I agree that it is not the most transparent implementation. I have put in place of this a cleaner method. Thanks!

**Comment by Utkarsh-Detha** [comment]
@jsiirola I have tried to follow the suggestions you made. For each suggestion that I implemented, I marked its corresponding comment as resolved. The comments that I did not mark as resolved are the ones that could use more of your attention. I was confused about:
- name change to 'mosek_direct' (can the meta-solver wait until another PR?)
- using the "_linear_canonical_form" attribute (any hints for an alternative approach)

Thank you for your immensely helpful review.

**Review by jsiirola — COMMENTED** [review]
@Utkarsh-Detha: Thank you for all your work on this.  I think we are very close.  As to your two outstanding questions, a simple wrapper class can be:
```python
@SolverFactory.register('mosek', doc='The MOSEK LP/MIP/QP solver')
class MOSEK(OptSolver):
    """The MOSEK LP/MIP/QP solver
    """

    def __new__(cls, *args, **kwds):
        mode = kwds.pop('solver_io', 'direct')
        #
        if mode in {'python', 'direct'}:
            opt = SolverFactory('mosek_direct', **kwds)
            if opt is None:
                logger.error('Python API for MOSEK is not installed')
            return opt
        if mode == 'persistent':
            opt = SolverFactory('mosek_persistent', **kwds)
            if opt is None:
                logger.error('Python API for GUROBI is not installed')
            return opt
        else:
            logger.error('Unknown/unsupported solver interface: %s' % (mode,))
            return None
```

As for using the "_linear_canonical_form" attribute, I don't have a good suggestion.  We should probably file this as an issue to consider when we finally get around to updating how we generate / manage canonical representation caches (and more generally, constraints that embed special storage structures (and not let that hold up this PR).

**Comment by jsiirola** [comment]
@michaelbynum: I think this PR is close, and we could really use your eyes on it (particularly around the persistent interface).

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:143]: In case of an invalid parameter name, an `AttributeError` will be raised when I use the `getattr` method to set the value for `param`. This exception will be caught in the except statement just below, and then re-raised. So this scenario should be covered.

**Review by jsiirola — APPROVED** [review]
This looks good to me.  Thank you @Utkarsh-Detha!

**Review by michaelbynum — CHANGES_REQUESTED** [review]
Overall, this looks good. I've requested a couple minor changes.

File: pyomo/opt/base/solvers.py
Comment (Utkarsh-Detha) [reply pyomo/opt/base/solvers.py:119]: Ah, you are right. I'll remove it.

File: pyomo/solvers/plugins/solvers/mosek_direct.py
Comment (Utkarsh-Detha) [reply pyomo/solvers/plugins/solvers/mosek_direct.py:290]: Agreed, I will make the change. 
However, I have a (sort-of) unrelated question about the wrapper method for adding constraints: if a user decides to add a cone, but one that is constructed as a domain, then they would be essentially adding a block. Do you think, then, that it would be a good idea to raise an error telling the user to switch to the add_block method, or would it be better to just call the add_block method from within the add_constraints wrapper method?

Comment (michaelbynum) [reply pyomo/solvers/plugins/solvers/mosek_direct.py]: That is an excellent question. I would lean toward calling add_block within the add_constraints wrapper, but I don't have a strong preference.

**Review by michaelbynum — APPROVED** [review]

**Comment by ghackebeil** [comment]
I would avoid calling add_block within add_constraint. I think that will just lead to confusion in the long term.

**Comment by michaelbynum** [comment]
Yeah, @ghackebeil is probably right...

**Review by jsiirola — APPROVED** [review]

**Comment by Utkarsh-Detha** [comment]
As the changes have been approved, I was wondering if there is something that needs to be done before the PR can be merged? Any hints would be appreciated. Thank you.

**Pull request merged**
